### PR TITLE
SALTO-6247 support mac_application in Jamf

### PR DESCRIPTION
**Pull request by omrifarkash**
On this PR:
- Fix Jamf build after upgrade to node 18
- Support fetch and deploy of `mac_application` type, behave as other type of Jamf classicApi
- Introduce new serialization strategy `idAndNameObject` to resolve a reference to `{ id: string, name: string }` as all Jamf classicAPI endpoints expect to recieve

---

---

_Release Notes_: 
_Jamf Adapter_:
- support mac_application fetch and deploy
- fix deploying references of site and category

---

_User Notifications_: 
_Jamf Adapter_:
- support mac_application fetch and deploy
- fix deploying references of site and category

### CODE REVIEW

File: packages/jamf-adapter/package.json [hunk 1]
```diff
@@ -17,7 +17,7 @@
   "main": "dist/index.js",
   "types": "dist/index.d.ts",
   "scripts": {
-    "build": "yarn concurrently \"yarn lint\" \"yarn build-ts\"",
+    "build": "../../build_utils/turbo_run.sh build-ts ; ../../build_utils/turbo_run.sh lint",
```
Comment (omrifarkash) [inline packages/jamf-adapter/package.json:20]: Fix Jamf build after upgrade to node 18. Confirmed by @eshedg

File: packages/jamf-adapter/src/definitions/deploy/classic_api_utils.ts [hunk 2]
```diff
@@ -42,7 +41,7 @@ export const createClassicApiDefinitionsForType = (
                 }
                 const parsedXML = xmljs.xml2js(value, { compact: true })
                 const id = _.get(parsedXML, `${typeName}.id._text`)
-                return { value: { id: shouldConvertIdToNumber ? Number(id) : id } }
+                return { value: { id: Number(id) } }
```
Comment (omrifarkash) [inline packages/jamf-adapter/src/definitions/deploy/classic_api_utils.ts:44]: Eventually I realized that all types need this id as number and not string

File: packages/jamf-adapter/src/definitions/fetch/transforms/utils.ts [hunk 3]
```diff
@@ -64,3 +64,13 @@ export const adjustServiceIdToTopLevel = (value: Record<string, unknown>): void
   _.set(general, 'id', undefined)
   value.id = id
 }
+
+/*
+ * Remove self_service_icon from self_service object
+ */
+export const removeSelfServiceIcon = (value: Record<string, unknown>): void => {
```
Comment (omrifarkash) [inline packages/jamf-adapter/src/definitions/fetch/transforms/utils.ts:71]: Removed this to utils as more than one type is using it now

File: packages/jamf-adapter/src/definitions/fetch/transforms/utils.ts [hunk 4]
```diff
@@ -64,3 +64,13 @@ export const adjustServiceIdToTopLevel = (value: Record<string, unknown>): void
...
+  const { self_service: selfService } = value
+  if (values.isPlainRecord(selfService)) {
+    delete selfService.self_service_icon
```
Comment (Uri-Bar) [inline packages/jamf-adapter/src/definitions/fetch/transforms/utils.ts:74]: 1. do we want to check if the deletion worked by its returned value?
2. do you want to move the uts of this func as well?

File: packages/jamf-adapter/src/definitions/references.ts [hunk 5]
```diff
@@ -64,6 +120,13 @@ const REFERENCE_RULES: referenceUtils.FieldReferenceDefinition<
...
-export const REFERENCES: definitions.ApiDefinitions['references'] = {
+export const REFERENCES: definitions.ApiDefinitions<Options>['references'] = {
   rules: REFERENCE_RULES,
+  serializationStrategyLookup: {
+    idAndNameObject: {
+      serialize: async ({ ref }) => ({ name: ref.value.value.name, id: ref.value.value.id }),
+      lookup: referenceUtils.basicLookUp,
+    },
+  },
+  fieldsToGroupBy: ['id'],
```
Comment (Uri-Bar) [inline packages/jamf-adapter/src/definitions/references.ts:131]: I think this is redundant as it always group by id

File: packages/jamf-adapter/src/definitions/references.ts [hunk 6]
```diff
@@ -64,6 +120,13 @@ const REFERENCE_RULES: referenceUtils.FieldReferenceDefinition<
   },
 ]
 
-export const REFERENCES: definitions.ApiDefinitions['references'] = {
+export const REFERENCES: definitions.ApiDefinitions<Options>['references'] = {
   rules: REFERENCE_RULES,
+  serializationStrategyLookup: {
```
Comment (Uri-Bar) [inline packages/jamf-adapter/src/definitions/references.ts:125]: cool

**Review by Uri-Bar — COMMENTED** [review]
looks good, just 2 small comments

File: packages/jamf-adapter/src/definitions/fetch/transforms/utils.ts (hunk 4 above)
Comment (omrifarkash) [reply packages/jamf-adapter/src/definitions/fetch/transforms/utils.ts:74]: 1. Don't think this is necessary.
2. Will do 👍

File: packages/jamf-adapter/src/definitions/references.ts (hunk 5 above)
Comment (omrifarkash) [reply packages/jamf-adapter/src/definitions/references.ts:131]: Tried to remove it. `fieldsToGroupBy` is mandatory if you define `serializationStrategyLookup`

Comment (Uri-Bar) [reply packages/jamf-adapter/src/definitions/references.ts:131]: I understand, I still dont think it nice.
just make sure you dont have any reference rule that uses the `name` field` as you override it

**Review by Uri-Bar — APPROVED** [review]

File: packages/jamf-adapter/src/definitions/references.ts (hunk 5 above)
Comment (omrifarkash) [reply packages/jamf-adapter/src/definitions/references.ts:131]: Nice, will change it to 'id' and 'name'

Comment (omrifarkash) [reply packages/jamf-adapter/src/definitions/references.ts:131]: On my next PR I do it

**Pull request merged**